Simplify positional arguments compatibility code in pytest.fixture()

bluetech opened the pull request:
The dynamic scope feature added in 10bf6aac76d5060a0db4a94871d6dcf0a1 necessitated some wrangling of arguments in pytest.fixture(). In particular, it deprecated positional arguments in favor of keyword-only arguments, while keeping backward compatibility.

The way it did this avoided some code duplication but ended up being quite hard to follow and to annotate with types.

Replace it with some straightforward code, which is not very DRY but is simple and easy to remove when the time comes.

<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1134,20 +1092,54 @@ def fixture(
+        warnings.warn(FIXTURE_POSITIONAL_ARGUMENTS, stacklevel=2)
+    # End backward compatiblity.
 
     if fixture_function and params is None and autouse is False:
```
BTW, I'm interested to know why the `params is None and autouse is False` condition exists here, if anyone knows... Seems weird to just ignore `fixture_function` if `params` or `autouse` are specified?

The-Compiler replied:
Judging from the `# direct decoration` comment, that's to support

```python
@pytest.fixture
def func():
    ....
```

rather than

```python
@pytest.fixture(...)
def func():
    ...
```

so there's nothing being ignored, because there are no arguments given at all in that case.

bluetech replied:
Right, the `if fixture_function` part I understand, what I'm having trouble figuring out is the `and params is None and autouse is False` part.

The reason I care about it is that I am working on adding type-annotating `pytest.fixture()`. Without this extra `params` condition, it is straightforward to distinguish the "direct decoration" case, but with it it is more difficult.

bluetech commented on src/_pytest/fixtures.py:
```diff
@@ -1134,20 +1092,54 @@ def fixture(
+    # End backward compatiblity.
 
     if fixture_function and params is None and autouse is False:
         # direct decoration
```
Also curious why `id` is omitted here (in the line below - GitHub doesn't let me comment on it).

bluetech replied:
Right. Still seems better to me not to ignore it, so any inconsistencies (like passing `ids` but not `params`) will be detected.

nicoddemus commented on src/_pytest/fixtures.py:
```diff
@@ -1031,48 +1031,6 @@ def __call__(self, function):
-FIXTURE_ARGS_ORDER = ("scope", "params", "autouse", "ids", "name")
-
-
-def _parse_fixture_args(callable_or_scope, *args, **kwargs):
```
TBH I'm OK with this implementation; I like that it is general enough that we can reuse it eventually when needed.

Having said that if others prefer the new version I'm OK with that too.

bluetech replied:
I don't think we would ever want to reuse it, because it's deprecated and new code should be keyword-only from the start. Any new parameters to `pytest.fixture()` would be keyword-only too. So this whole code block is "read-only" and I think it's better to have it straightforward and localized (and shorter).

nicoddemus replied:
Yeah I have that feeling like I wanted to reuse this somewhere, but can't really remember it right now. 😁

bluetech commented:
After looking again I've made some further changes:

- Renamed `callable_or_scope` to `fixture_function`. The previous name is really internal-facing, let's use the proper name we would use without compat/deprecation. Also moved its compat check to inside the block marked with the comments.

- Moved `if params is not None: params = list(params)` to *after* the compat check -- was a slight bug in the previous code.

- Added tests to fix the coverage hopefully.

bluetech commented:
> It kinda hurts my eyes seeing so much if statements and makes me want to immediately revert to the previous code (as I'm the original author). But I need to agree, that this version is much easier to read and remove in the future.

:smile: Definitely ugly, but it's localized and easier to handle when adding type annotations, so I think the advantages outweigh the ugliness.
